presto-main/src/main/java/com/facebook/presto/sql/planner/PlanFragmenter.java

```diff
@@ -362,4 +383,85 @@ public PartitioningHandle getPartitioningHandle()
             return partitionedSources;
         }
     }
+
+    private static class GroupedExecutionTagger
+            extends PlanVisitor<GroupedExecutionProperties, Void>
+    {
+        private final Session session;
+        private final Metadata metadata;
+        private final NodePartitioningManager nodePartitioningManager;
+
+        public GroupedExecutionTagger(Session session, Metadata metadata, NodePartitioningManager nodePartitioningManager)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+            this.nodePartitioningManager = requireNonNull(nodePartitioningManager, "nodePartitioningManager is null");
+        }
+
+        @Override
+        protected GroupedExecutionProperties visitPlan(PlanNode node, Void context)
+        {
+            if (node.getSources().isEmpty()) {
+                return new GroupedExecutionProperties(false, false);
+            }
+            return processChildren(node);
+        }
+
+        @Override
+        public GroupedExecutionProperties visitJoin(JoinNode node, Void context)
+        {
+            GroupedExecutionProperties properties = processChildren(node);
```

Comment:
Please add a comment on why we don't need to check if the distributions are the same.

Reply:
I added requested comments to `processChildren`, where `isCurrentNodeCapable` is derived. Determination of whether the current can take advantage of grouped exeuction is what depends on children being compatible.